saleor/graphql/checkout/mutations.py

```diff
@@ -193,23 +215,36 @@ def clean_input(cls, info, instance, data):
         return cleaned_input
 
     @classmethod
-    @transaction.atomic()
-    def save(cls, info, instance, cleaned_input):
+    def save_addresses(
+        cls,
+        instance: models.Checkout,
+        cleaned_input: dict,
+        variants: List[product_models.ProductVariant],
+    ):
         shipping_address = cleaned_input.get("shipping_address")
         billing_address = cleaned_input.get("billing_address")
 
-        if shipping_address:
+        if shipping_address and any(
+            variant.is_shipping_required() for variant in variants
```

Comment:
What about `checkout.is_shipping_required()`? Could we use it here?

Reply:
The issue with this is that there are no lines yet in the checkout object. We could add the lines first, but then it would add an additional query as we would have to:
1. create checkout
1. add the lines
1. add the addresses
1. save checkout

instead of
1. add the addresses
1. create the checkout
1. add the lines

Reply:
I'm only worried that if we changed the logic of `checkout.is_shipping_required` function for some reason, this is one more place in the code that needs to be updated as well. I think one additional query when saving is not a big deal, it's still O(1).